### PR TITLE
[Truffle] Implementing more Array#[]=.

### PR DESCRIPTION
**Pull request by bjfish**

### CODE REVIEW

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java [hunk 1]
```diff
@@ -423,29 +423,71 @@ public IndexSetNode(IndexSetNode prev) {
...
             popNode = prev.popNode;
+            toIntNode = prev.toIntNode;
         }
 
         @Specialization
         public Object set(VirtualFrame frame, RubyArray array, int index, Object value, UndefinedPlaceholder unused) {
+            final int normalizedIndex = array.normalizeIndex(index);
+            if (normalizedIndex < 0) {
+                CompilerDirectives.transferToInterpreter();
+                String errMessage = "index " + index + " too small for array; minimum: " + Integer.toString(-array.getSize());
```
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java]: This is what MRI says, right?

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java [hunk 2]
```diff
@@ -557,16 +658,30 @@ public Object setOtherArray(VirtualFrame frame, RubyArray array, int start, int
...
+            }
             return setObject(frame, array, normalizedStart, length, other);
         }
 
         @Specialization(guards = "!areBothIntegerFixnum")
         public Object setRangeArray(VirtualFrame frame, RubyArray array, RubyRange.IntegerFixnumRange range, RubyArray other, UndefinedPlaceholder unused) {
             final int normalizedStart = array.normalizeIndex(range.getBegin());
-            final int normalizedEnd = range.doesExcludeEnd() ? array.normalizeIndex(range.getEnd()) - 1 : array.normalizeIndex(range.getEnd());
+            int normalizedEnd = range.doesExcludeEnd() ? array.normalizeIndex(range.getEnd()) - 1 : array.normalizeIndex(range.getEnd());
+            if( normalizedEnd < 0 ) {
```
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java]: Spacing, please

Comment (bjfish) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java]: I've updated the spacing

File: truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java (hunk 1 above)
Comment (bjfish) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/ArrayNodes.java]: Yes

**Pull request merged**